sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala

```diff
@@ -42,15 +42,55 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks {
     InternalRow.fromSeq(values.map(CatalystTypeConverters.convertToCatalyst))
   }
 
+  protected def convertToCatalystUnsafe(a: Any): Any = a match {
+    case arr: Array[Boolean] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Byte] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Short] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Int] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Long] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Float] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Double] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case other => CatalystTypeConverters.convertToCatalyst(other)
+  }
+
   protected def checkEvaluation(
       expression: => Expression, expected: Any, inputRow: InternalRow = EmptyRow): Unit = {
     val serializer = new JavaSerializer(new SparkConf()).newInstance
     val expr: Expression = serializer.deserialize(serializer.serialize(expression))
-    val catalystValue = CatalystTypeConverters.convertToCatalyst(expected)
+    // No codegen version expects GenericArrayData
+    val catalystValue = expected match {
```

Comment:
At the callee side, we can do type aware comparison so that the caller side can just pass in a generic array data, e.g. https://github.com/apache/spark/pull/15645/files#diff-41747ec3f56901eb7bfb95d2a217e94dR69